# 2. Exploring Data

## 2.0. Import Necessary Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

This section imports the essential Python libraries required for data analysis, visualization, and statistical exploration. Each library serves a specific purpose:

- **NumPy (np)**: Provides support for multi-dimensional arrays and a wide range of mathematical operations, facilitating efficient numerical computations.
- **Pandas (pd)**: A powerful data manipulation library used for loading, cleaning, and analyzing structured data through its **DataFrame** and **Series** objects.
- **Matplotlib (plt)**: A versatile library for creating static, interactive, and animated visualizations, primarily used for plotting data in 2D.
- **Seaborn (sns)**: Built on top of Matplotlib, this library simplifies the creation of visually appealing and informative statistical graphics, offering a high-level interface for data visualization.

These libraries together enable efficient data processing and insightful visual representations to support data-driven decision-making.

## 2.1. Read data

In [2]:
data_path = './Data/CarInformation.csv'
df = pd.read_csv(data_path)

In this section, we use the **Pandas** library to read data from a CSV file containing information about automobiles. 

- First, the file path is stored in the variable `data_path`.
- Then, the `pd.read_csv()` function is called to load the data from the file and convert it into a **DataFrame** object, which allows for easier manipulation, analysis, and visualization of the data.
- The `df` variable holds the loaded dataset, which can be used for further processing in the subsequent steps of the analysis.

This approach simplifies data handling and prepares the dataset for deeper exploration and analysis.

## 2.2. How many rows and how many columns?

In [3]:
row, col = df.shape
print("Number of Row: ", row)
print("Number of Column: ", col)

Number of Row:  398
Number of Column:  9


The dataset contains **398 rows** and **9 columns**. This indicates that there are **398 entries** or **records** in the dataset, each corresponding to a different **automobile**. The **9 columns** represent various attributes of these automobiles, which might include characteristics such as **fuel efficiency**, **engine specifications**, and **model details**.

With this structure, the dataset provides a **comprehensive set of features** for each vehicle, offering sufficient data for **analysis** and **model building**. The relatively moderate number of rows suggests that while the dataset is **not very large**, it still provides a **meaningful sample** for **analysis** and **pattern discovery**.

## 2.3. What is the meaning of each row?

In [4]:
df.head()

name   mpg  cylinders  displacement  horsepower  \
0  chevrolet chevelle malibu  18.0          8         307.0       130.0   
1          buick skylark 320  15.0          8         350.0       165.0   
2         plymouth satellite  18.0          8         318.0       150.0   
3              amc rebel sst  16.0          8         304.0       150.0   
4                ford torino  17.0          8         302.0       140.0   

   weight  acceleration  model_year origin  
0    3504          12.0          70    usa  
1    3693          11.5          70    usa  
2    3436          11.0          70    usa  
3    3433          12.0          70    usa  
4    3449          10.5          70    usa

Each **row** in the dataset represents the **information of a specific automobile**. Each row contains a set of **attributes** related to that particular vehicle, which describe its **characteristics** and **performance**. These rows collectively provide a **detailed profile** for every car, allowing for **comparisons** across different vehicles based on various features such as **fuel efficiency**, **engine specifications**, and **model year**.

Essentially, every row captures the **key data** for an individual car, and this **structured representation** helps in understanding **patterns**, **trends**, and **relationships** within the dataset. By examining the rows, we can analyze how different cars vary based on their **attributes** and make **informed decisions** or **predictions**.

## 2.4. Are there duplicated rows?

In [5]:
df[df.duplicated(keep=False)]

Empty DataFrame
Columns: [name, mpg, cylinders, displacement, horsepower, weight, acceleration, model_year, origin]
Index: []

Upon checking the dataset for **duplicated rows** using the command `df[df.duplicated(keep=False)]`, it is observed that there are **no duplicated rows** present. This means that each **entry** in the dataset is **unique**, which is a good sign as it ensures that there is no **redundancy** in the data. 

The **absence of duplicate rows** is important because it prevents potential **biases** or **errors** that could arise from analyzing repeated data. It suggests that the data has been **properly collected** and **cleaned**, making it suitable for further **analysis** and **modeling** without the need for **data cleansing** regarding duplicates.

## 2.5. What is the meaning of each column?

In [6]:
list(df.columns.values)

['name',
 'mpg',
 'cylinders',
 'displacement',
 'horsepower',
 'weight',
 'acceleration',
 'model_year',
 'origin']

Here is a detailed explanation of each column in the dataset:

| **Column**     | **Description**                                                                 | **Meaning** |
|----------------|---------------------------------------------------------------------------------|-------------|
| **name**       | Unique identifier for each car.                                                 | This is the unique identifier for each car, usually representing the car's model or specific name. It is used to differentiate between individual cars in the dataset, providing a reference point for each vehicle. |
| **mpg**        | Fuel efficiency measured in miles per gallon. 1 Gallon = 3,785411784 liters      | MPG stands for "Miles per Gallon," which measures the fuel efficiency of the car. It indicates how far the car can travel on one gallon of fuel. A higher MPG value typically reflects better fuel economy, which is a key factor for buyers looking to reduce fuel costs and minimize their environmental impact. 1 Gallon is equivalent to 3.785411784 liters. |
| **cylinders**  | Number of cylinders in the engine.                                              | This column represents the number of cylinders in the engine. The number of cylinders affects the engine's power output and efficiency. Cars with more cylinders typically have more power, but they may also consume more fuel, while cars with fewer cylinders tend to be more fuel-efficient. |
| **displacement**| Engine displacement, indicating its size or capacity.                            | Engine displacement refers to the total volume of all the cylinders in an engine, usually measured in liters or cubic inches. It gives an indication of the engine’s size and its potential to produce power. Larger displacement engines generally provide more horsepower and torque but may be less fuel-efficient. |
| **horsepower** | Power output of the engine.                                                     | Horsepower indicates the power output of the car’s engine. It is a measure of the engine's ability to do work, with higher values representing more powerful engines that can achieve faster speeds or handle more strenuous tasks, such as towing. |
| **weight**     | Weight of the car.                                                              | The weight of the car is listed in this column. Heavier cars tend to have more stability on the road, but they can be less fuel-efficient and slower to accelerate. The weight also plays a role in the car's safety, as heavier vehicles typically provide better protection in collisions. |
| **acceleration**| Capability to increase speed, measured in seconds.                              | This column measures the car's ability to accelerate, typically represented by the time it takes to go from 0 to 60 miles per hour (0-100 km/h). Cars with faster acceleration times are generally considered more performance-oriented and can offer a more dynamic driving experience. |
| **model_year** | Year of manufacture for the car model.                                          | The model year of the car refers to the year the car was manufactured or the year the model was introduced to the market. This column helps identify the age of the vehicle, which can affect its value, technology, and reliability. |
| **origin**     | Country or region of origin for each car.                                       | The origin column indicates the country or region where the car was manufactured or designed. This can provide insights into the car's design philosophy, quality standards, and the reputation of manufacturers from specific regions (e.g., American, European, or Japanese cars). |


These **columns** represent the **key characteristics** that help describe each **automobile** in the dataset. Understanding these **attributes** is important for **analyzing** the cars, **comparing** their **performance**, and making **data-driven decisions**.

## 2.6. What is the current data type of each column? Are there columns having inappropriate data types?

### 2.6.1. Overview of Data Types:

To get a **general overview** of the dataset without detailed analysis or statistical insights, you can simply display the **data types** of each **column** and a quick **summary** of each column (e.g., **unique values** for categorical columns and **summary statistics** for numerical columns).

In [7]:
df.dtypes

name             object
mpg             float64
cylinders         int64
displacement    float64
horsepower      float64
weight            int64
acceleration    float64
model_year        int64
origin           object
dtype: object

The **data types** of each **column** in the dataset provide useful insights into the **structure** of the data. Here’s a breakdown and interpretation of each column's **data type**:

| **Column Name**  | **Data Type** | **Description** |
|------------------|---------------|-----------------|
| **name**         | object        | This column is an identifier for each car, and it is of type `object`, which typically means that the data is stored as text or strings. This is appropriate for a column that holds car names or unique identifiers. |
| **mpg**          | float64       | The `mpg` (miles per gallon) column is of type `float64`, which indicates that it holds numerical data with decimals. This is appropriate, as fuel efficiency is usually represented with decimal numbers. |
| **cylinders**    | int64         | The `cylinders` column is of type `int64`, which means it stores integer values. This is expected since the number of cylinders in an engine is a whole number. |
| **displacement** | float64       | The `displacement` column is of type `float64`, which is appropriate because engine displacement can have decimal values (e.g., 2.5 liters). |
| **horsepower**   | float64       | The `horsepower` column is of type `float64`, meaning it stores numerical values with decimals. This makes sense since engine power can vary and be expressed with fractional numbers. |
| **weight**       | int64         | The `weight` column is of type `int64`, which indicates that the weight of the car is represented as an integer. This is correct, as weight is usually measured in whole units (e.g., kilograms or pounds). |
| **acceleration** | float64       | The `acceleration` column is of type `float64`, which makes sense as acceleration times are typically measured with decimal precision (e.g., 8.5 seconds to 60 mph). |
| **model_year**   | int64         | The `model_year` column is of type `int64`, which is appropriate since model years are typically represented by whole integers (e.g., 1985, 2020). |
| **origin**       | object        | The `origin` column is of type `object`, meaning it holds text data (typically the country or region of origin). This is correct as the origin is generally a categorical variable represented by names (e.g., 'USA', 'Europe', 'Japan'). |

#### Summary:
- **_Correct Data Types_**: Most **columns** have the appropriate **data types**. **_Numerical columns_** like **`mpg`**, **`horsepower`**, **`displacement`**, **`acceleration`**, etc., are correctly represented as **`float64`**, while whole numbers like **`cylinders`**, **`weight`**, and **`model_year`** are appropriately stored as **`int64`**.
- **_Categorical Columns_**: **`name`** and **`origin`** are stored as **`object`**, which is correct for **_textual_** or **_categorical data_**.


#### **_Potential Improvements_**:
- **_Missing Data Check_**: It’s important to check if there are any missing values in the dataset, especially for numerical columns like **`horsepower`**, which could sometimes have missing or incorrect values.
- **_Consistency in `origin` Column_**: If **`origin`** has categories like **`'USA'`**, **`'Europe'`**, and **`'Japan'`**, it might be beneficial to convert it into a **categorical** data type (**`category`**) for efficiency and consistency in processing.

Overall, the data types are mostly appropriate for the nature of the data in each column.

### 2.6.2. Quick Summary of Columns:

For a brief summary of the columns (showing the number of **unique values**, **null values**, etc.), you can use **`info()`**:

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   name          398 non-null    object 
 1   mpg           398 non-null    float64
 2   cylinders     398 non-null    int64  
 3   displacement  398 non-null    float64
 4   horsepower    392 non-null    float64
 5   weight        398 non-null    int64  
 6   acceleration  398 non-null    float64
 7   model_year    398 non-null    int64  
 8   origin        398 non-null    object 
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


The DataFrame consists of **398 entries** and **9 columns**. Below is a detailed summary of the columns:

| **Column Name**   | **Non-Null Count** | **Data Type** | **Notes** |
|-------------------|--------------------|---------------|-----------|
| **name**          | 398 non-null       | object        | No missing values, this column stores the car's name or identifier. |
| **mpg**           | 398 non-null       | float64       | No missing values, stores miles per gallon, a numerical column. |
| **cylinders**     | 398 non-null       | int64         | No missing values, stores the number of cylinders in the car's engine. |
| **displacement**  | 398 non-null       | float64       | No missing values, stores engine displacement in liters, a numerical value. |
| **horsepower**    | 392 non-null       | float64       | There are 6 missing values, stores the car's horsepower, a numerical value. |
| **weight**        | 398 non-null       | int64         | No missing values, stores the car's weight in kilograms or pounds. |
| **acceleration**  | 398 non-null       | float64       | No missing values, stores acceleration time (e.g., 0-60 mph) as a decimal value. |
| **model_year**    | 398 non-null       | int64         | No missing values, stores the model year of the car. |
| **origin**        | 398 non-null       | object        | No missing values, stores the car's country of origin (categorical). |

#### Key Observations:
- The DataFrame is mostly complete with no missing values in most columns except for **horsepower**, which has 6 missing entries.
- The columns contain a mix of numerical (`int64`, `float64`) and categorical (`object`) data types.
- The memory usage is **28.1 KB**.

#### Recommendations:
- The missing values in the **horsepower** column should be addressed (e.g., imputation, removal) based on further analysis.

### 2.6.3. Quick Statistics for Numerical Columns:

For numerical columns, you can get a basic summary with describe():

In [9]:
df.describe()

mpg   cylinders  displacement  horsepower       weight  \
count  398.000000  398.000000    398.000000  392.000000   398.000000   
mean    23.514573    5.454774    193.425879  104.469388  2970.424623   
std      7.815984    1.701004    104.269838   38.491160   846.841774   
min      9.000000    3.000000     68.000000   46.000000  1613.000000   
25%     17.500000    4.000000    104.250000   75.000000  2223.750000   
50%     23.000000    4.000000    148.500000   93.500000  2803.500000   
75%     29.000000    8.000000    262.000000  126.000000  3608.000000   
max     46.600000    8.000000    455.000000  230.000000  5140.000000   

       acceleration  model_year  
count    398.000000  398.000000  
mean      15.568090   76.010050  
std        2.757689    3.697627  
min        8.000000   70.000000  
25%       13.825000   73.000000  
50%       15.500000   76.000000  
75%       17.175000   79.000000  
max       24.800000   82.000000

#### Key Observations

| **Observation**                | **Details**                                                                 |
|---------------------------------|-----------------------------------------------------------------------------|
| **mpg (Miles per Gallon)**     | Mean is **23.51**, with a wide range (**9.00** to **46.60**), indicating variability in fuel efficiency across cars. |
| **cylinders**                   | Most cars have either **4** or **8** cylinders. The mean is **5.45**, with a range from **3** to **8**. |
| **displacement**                | The average displacement is **193.43**, with values ranging from **68.00** to **455.00**, indicating varying engine sizes. |
| **horsepower**                  | Average horsepower is **104.47**, ranging from **46** to **230**. The variability in horsepower is high, as indicated by the standard deviation of **38.49**. |
| **weight**                      | Mean weight is **2970.42**, ranging from **1613.00** to **5140.00**, showing significant differences in car sizes. |
| **acceleration**                | The average acceleration time is **15.57** seconds, ranging from **8.00** to **24.80** seconds. Moderate variation in acceleration performance is observed. |
| **model_year**                  | The model years range from **70** to **82**, with a mean of **76.01**, indicating most cars are from the late 1970s and early 1980s. |


#### Recommendations:
- The data shows variability in many columns, which might suggest that different car types (e.g., sedans, sports cars, etc.) are present.
- The missing data in the **horsepower** column should be addressed, possibly through imputation or analysis based on other features.

#### Handling the `model_year` Column

The `model_year` column initially contains values representing the year of the car model, but these values are in a two-digit format (e.g., '70' for 1970). To convert these values into a more standard four-digit year format (e.g., '1970'), we will add 1900 to each value.

##### Step 1: Inspect the Maximum Value of `model_year`

To understand the current state of the `model_year` column, we first check the maximum value.

In [10]:
print(df['model_year'].max())

82


This will output the maximum value in the column, which is 82 in the dataset, representing the model year '82' (or 1982).

##### Step 2: Convert to Four-Digit Year Format

Next, we add 1900 to each value in the model_year column to convert it to the four-digit year format:

In [11]:
df['model_year'] = df['model_year'] + 1900

After this operation, the model_year column will contain values like 1970, 1982, etc., instead of 70, 82, making the data more consistent and easier to interpret.

##### Step 3: Final Check

To confirm the transformation, we can print the updated model_year column:

In [12]:
df['model_year']

0      1970
1      1970
2      1970
3      1970
4      1970
       ... 
393    1982
394    1982
395    1982
396    1982
397    1982
Name: model_year, Length: 398, dtype: int64

This will display the column with the corrected four-digit years.